content/transfer-hook.md

```diff
@@ -0,0 +1,2918 @@
+
+#[derive(Accounts)]
+
+pub struct TransferHook<'info> {
+
+  #[account(token::mint = mint, token::authority = owner)]
+
+  pub source_token: InterfaceAccount<'info, TokenAccount>,
+
+  pub mint: InterfaceAccount<'info, Mint>,
+
+  #[account(token::mint = mint)]
+
+  pub destination_token: InterfaceAccount<'info, TokenAccount>,
+
+  /// CHECK: source token account owner
+
+  pub owner: UncheckedAccount<'info>,
+
+
+
+  /// CHECK: ExtraAccountMetaList Account,
+
+  #[account(seeds = [b"extra-account-metas", mint.key().as_ref()], bump)]
+
+  pub extra_account_meta_list: UncheckedAccount<'info>,
+
+
+
+  pub token_program: Interface<'info, TokenInterface>,
+
+
+
+  pub associated_token_program: Program<'info, AssociatedToken>,
+
+
+
+  pub crumb_mint: InterfaceAccount<'info, Mint>,
+
+
+
+  /// CHECK: mint authority Account,
+
+  #[account(seeds = [b"mint-authority"], bump)]
+
+  pub mint_authority: UncheckedAccount<'info>,
+
+
+
+  #[account(
+
+        token::mint = crumb_mint,
+
+        token::authority = owner,
+
+    )]
+
+  pub crumb_mint_ata: InterfaceAccount<'info, TokenAccount>,
+
+}
+
+```
+
+
+
+**`transfer_hook` Instruction**
+
+
+
+This instruction is fairly simple, it will only make one CPI to the Token Program to mint a new crumb token for each transfer, all that we need to do is to pass the right accounts to the `mint_to` CPI.
+
+
+
+Since the mint_authority is a PDA of the transfer hook program itself, the program can sign for it. Therefore we'll use `new_with_signer` and pass mint_authority seeds as the signer seeds.
+
+
+
+```rust
+
+pub fn transfer_hook(ctx: Context<TransferHook>, _amount: u64) -> Result<()> {
+
+  let signer_seeds: &[&[&[u8]]] = &[&[b"mint-authority", &[ctx.bumps.mint_authority]]];
+
+  // mint a crumb token for each transaction
+
+  token::mint_to(
+
+    CpiContext::new_with_signer(
+
+      ctx.accounts.token_program.to_account_info(),
+
+      token::MintTo {
+
+        mint: ctx.accounts.crumb_mint.to_account_info(),
+
+        to: ctx.accounts.crumb_mint_ata.to_account_info(),
+
+        authority: ctx.accounts.mint_authority.to_account_info(),
+
+      },
+
+      signer_seeds
+
+    ),
+
+    1
+
+  ).unwrap();
+
+
+
+  Ok(())
+
+}
+
+```
+
+
+
+You may have noticed that we are using `token::mint_to` instead of `token_2022::mint_to`, additionally in the `extra_account_meta_list` we're saving the Token Program, not the Token Extensions Program. This is because the crumb SFT *has* to be a Token Program mint and not a Token Extensions Program mint. The answer as to why, is interesting. When first writing this, we wanted to make both the Cookie and Crumb tokens to be Token Extensions Program mints. However, when we did this, we would get a very interesting error: `No Reentrancy`. This happens because the transfer hook is called as a CPI from within the Token Extensions Program, and Solana does not allow [recursive CPIs into the same program](https://defisec.info/solana_top_vulnerabilities). 
```

Comment:
```suggestion
You may have noticed that we are using `token::mint_to` instead of `token_2022::mint_to`, additionally in the `extra_account_meta_list` we're saving the Token Program, not the Token Extensions Program. This is because the crumb SFT *has* to be a Token Program mint, not a Token Extensions Program mint. The reason why is interesting: when first writing this, we wanted to make both the Cookie and Crumb tokens to be Token Extensions Program mints. However, when we did this, we would get a very interesting error: `No Reentrancy`. This happens because the transfer hook is called as a CPI from within the Token Extensions Program, and Solana does not allow [recursive CPIs into the same program](https://defisec.info/solana_top_vulnerabilities). 
```

Reply:
Fixed in latest pushes